pkg/fileutils/fileutils_unix.go

```diff
@@ -6,17 +6,93 @@ import (
 	"context"
 	"fmt"
 	"os"
+	"unsafe"
 
 	"github.com/containerd/containerd/log"
+	"golang.org/x/sys/unix"
 )
 
 // GetTotalUsedFds Returns the number of used File Descriptors by
 // reading it via /proc filesystem.
 func GetTotalUsedFds() int {
-	if fds, err := os.ReadDir(fmt.Sprintf("/proc/%d/fd", os.Getpid())); err != nil {
-		log.G(context.TODO()).Errorf("Error opening /proc/%d/fd: %s", os.Getpid(), err)
-	} else {
-		return len(fds)
+	nameCount, err := readDirnamesCount(fmt.Sprintf("/proc/%d/fd", os.Getpid()))
```

Comment:
From https://docs.kernel.org/filesystems/proc.html#proc-pid-fd-list-of-symlinks-to-open-files:

> The number of open files for the process is stored in 'size' member of stat() output for /proc/<pid>/fd for fast access.

Reply:
Per @corhere; only on kernel 6.4+. We should use this as a fast path, and fall back to `readdir(2)`

Reply:
For future reference: the fast path was added to the kernel in https://github.com/torvalds/linux/commit/f1f1f2569901ec5b9d425f2e91c09a0e320768f3